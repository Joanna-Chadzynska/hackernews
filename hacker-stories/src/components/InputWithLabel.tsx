import React, { ReactNode } from "react";
import styled from "styled-components";

const StyledLabel = styled.label`
	border-top: 1px solid #171212;
	border-left: 1px solid #171212;
	font-size: 1.5rem;
	padding-left: 5px;
`;

const StyledInput = styled.input`
	border: none;
	border-bottom: 1px solid #171212;
	background-color: transparent;
	font-size: 1.5rem;
`;

export interface InputWithLabelProps {
	id: string;
	value: string;
	onInputChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
	type?: string;
	children?: ReactNode;
	isFocused?: boolean;
}
const InputWithLabel: React.SFC<InputWithLabelProps> = ({
	id,
	type = "text",
	value,
	onInputChange,
	children,
	isFocused
}) => {
	const inputRef = React.useRef<HTMLInputElement>(null);

	React.useEffect(() => {
		if (isFocused && inputRef.current) {
			inputRef.current.focus();
		}
	}, [isFocused]);

	return (
		<>
			<StyledLabel htmlFor={id}>{children}</StyledLabel>
			&nbsp;
			<StyledInput
				ref={inputRef}
				id={id}
				type={type}
				value={value}
				onChange={onInputChange}
				autoFocus={isFocused}
			/>
		</>
	);
};

export default InputWithLabel;
