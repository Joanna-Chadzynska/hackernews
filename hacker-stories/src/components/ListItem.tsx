import React from "react";

export interface ListItemProps {
	title: string;
	url: string;
	author: string;
	num_comments: number;
	points: number;
	objectID: number;
}

const ListItem: React.SFC<ListItemProps> = ({
	title,
	url,
	author,
	num_comments,
	points
}) => {
	return (
		<div>
			<span>
				<a href={url}>{title}</a>
			</span>
			<span>{author}</span>
			<span>{num_comments}</span>
			<span>{points}</span>
		</div>
	);
};

export default ListItem;
