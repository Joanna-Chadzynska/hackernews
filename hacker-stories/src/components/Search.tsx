import React from "react";
import InputWithLabel from "./InputWithLabel";

export interface SearchProps {
	onSearch: any;
	search: string;
}

const Search: React.SFC<SearchProps> = ({ onSearch, search }) => {
	return (
		<div>
			{/* <InputWithLabel
				id='search'
				name='search'
				type='text'
				value={search}
				onInputChange={onSearch}
			/> */}
		</div>
	);
};

export default Search;
