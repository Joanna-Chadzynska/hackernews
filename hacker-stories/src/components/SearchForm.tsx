import * as React from "react";
import InputWithLabel from "./InputWithLabel";
import styles from "../App.module.css";

export interface SearchFormProps {
	searchTerm: string;
	onSearchInput: any;
	onSearchSubmit: any;
}

const SearchForm: React.SFC<SearchFormProps> = ({
	searchTerm,
	onSearchInput,
	onSearchSubmit
}) => {
	return (
		<form onSubmit={onSearchSubmit} className={styles.searchForm}>
			<InputWithLabel
				id='search'
				value={searchTerm}
				onInputChange={onSearchInput}
				isFocused>
				<strong>Search: </strong>
			</InputWithLabel>
			<button
				type='submit'
				disabled={!searchTerm}
				className={`${styles.button} ${styles.buttonLarge}`}>
				Submit
			</button>
		</form>
	);
};

export default SearchForm;
