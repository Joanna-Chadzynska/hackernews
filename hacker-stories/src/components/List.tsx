import * as React from "react";
import ListItem from "./ListItem";

export interface ListProps {
	list: any;
	onRemoveItem: any;
}

const List: React.SFC<ListProps> = ({ list, onRemoveItem }) =>
	list.map((item: any) => (
		<ListItem key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
	));

export default List;
