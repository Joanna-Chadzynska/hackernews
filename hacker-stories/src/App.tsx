import React, { useEffect, useReducer, useCallback, useState } from "react";
// import { Counter } from "./features/counter/Counter";
// import styles from "./App.module.css";
import List from "./components/List";
import useSemiPersistentState from "./customHooks/index";
import axios from "axios";
import { ReactComponent as Check } from "./assets/logo.svg";
import styled from "styled-components";
import SearchForm from "./components/SearchForm";

export interface AppProps {}

const StyledContainer = styled.div`
	height: 100vw;
	padding: 1.25rem;
	background: #83a4d4;
	background: linear-gradient(to left, #b6fbff, #83a4d4);
	color: #171212;
`;

const StyledHeadlinePrimary = styled.h1`
	font-size: 3rem;
	font-weight: 300;
	letter-spacing: 2px;
`;

const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

export const storiesReducer = (state: any, action: any) => {
	switch (action.type) {
		case "STORIES_FETCH_INIT":
			return {
				...state,
				isLoading: true,
				isError: false
			};

		case "STORIES_FETCH_SUCCESS":
			return {
				...state,
				isLoading: false,
				isError: false,
				data: action.payload
			};

		case "STORIES_FETCH_FAILURE":
			return {
				...state,
				isLoading: false,
				isError: true
			};

		case "REMOVE_STORY":
			return {
				...state,
				data: state.data.filter(
					(story: any) => action.payload.objectID !== story.objectID
				)
			};

		default:
			throw new Error();
	}
};

// const getAsyncStories = () =>
// 	new Promise((resolve) =>
// 		setTimeout(() => resolve({ data: { stories: initialStories } }), 2000)
// 	);

const App: React.SFC<AppProps> = () => {
	const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");

	const [url, setUrl] = useState(`${API_ENDPOINT}${searchTerm}`);

	const [stories, dispatchStories] = useReducer(storiesReducer, {
		data: [],
		isLoading: false,
		isError: false
	});

	const handleFetchStories = useCallback(async () => {
		if (!searchTerm) return;
		dispatchStories({ type: "STORIES_FETCH_INIT" });

		try {
			const result = await axios.get(url);

			dispatchStories({
				type: "STORIES_FETCH_SUCCESS",
				payload: result.data.hits
			});
		} catch {
			dispatchStories({ type: "STORES_FETCH_FAILURE" });
		}
	}, [url]);

	useEffect(() => {
		handleFetchStories();
	}, [handleFetchStories]);

	const handleRemoveStory = (item: any) => {
		dispatchStories({
			type: "REMOVE_STORY",
			payload: item
		});
	};

	// const getTitle = (title: string) => title;

	const handleSearchInput = (e: any) => {
		setSearchTerm(e.target.value);
	};

	const handleSearchSubmit = (e: any) => {
		e.preventDefault();
		setUrl(`${API_ENDPOINT}${searchTerm}`);
	};

	// const searchedStories = stories.data.filter((story: any) =>
	// 	story.title.toLowerCase().includes(searchTerm.toLowerCase())
	// );

	return (
		<StyledContainer>
			{/* <h1>Hello, {getTitle("React with typeScript")}</h1> */}

			<StyledHeadlinePrimary>My Hacker Stories</StyledHeadlinePrimary>

			<SearchForm
				searchTerm={searchTerm}
				onSearchInput={handleSearchInput}
				onSearchSubmit={handleSearchSubmit}
			/>

			{/* <header className='App-header'>
				<Counter />
			</header> */}
			<hr />

			{stories.isError && <p>Something went wrong ...</p>}

			{stories.isLoading ? (
				<p>Loading...</p>
			) : (
				<List list={stories.data} onRemoveItem={handleRemoveStory} />
			)}
		</StyledContainer>
	);
};

export default App;
