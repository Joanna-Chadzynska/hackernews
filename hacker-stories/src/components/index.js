import ListItem from "./+List/ListItem";
import List from "./+List";
import SearchForm from "./+SearchForm";
import InputWithLabel from "./+InputWithLabel";
import Messages from "./+Messages";
import Users from "./+Users";

export { ListItem, List, SearchForm, InputWithLabel, Messages, Users };
