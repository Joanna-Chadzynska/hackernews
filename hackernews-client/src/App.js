import React, { Component } from 'react';
import axios from 'axios';
import { Table, Search } from './components';
import { DEFAULT_QUERY, searchUrl } from './utils/api';
import './App.css';
import Button from './components/Button';

// const App = () => {
//   const [searchTerm, setSearchTerm] = React.useState('');

//   const handleChange = event => {
//     setSearchTerm(event.target.value);
//   };

//   return (
//     <div>
//       <h1>My Hacker Stories</h1>

//       <label htmlFor="search">Search: </label>
//       <input id="search" type="text" onChange={handleChange} />

//       <p>
//         Searching for <strong>{searchTerm}</strong>.
//       </p>

//       <hr />

//       <List list={stories} />
//     </div>
//   );
// };

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
    };
  }

  needsToSearchTopStories = searchTerm => {
    return !this.state.results[searchTerm];
  };

  onDismiss = id => {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page },
      },
    });
  };

  onChangeSearch = e => {
    this.setState({ searchTerm: e.target.value });
  };

  setSearchTopStories = result => {
    const { hits, page } = result;
    const { searchKey, results } = this.state;

    const oldHits =
      results && results[searchKey] ? results[searchKey].hits : [];

    const updatedHits = [...oldHits, ...hits];

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page },
      },
    });
  };

  fetchSearchTopStories = (searchTerm, page = 0) => {
    fetch(searchUrl(searchTerm, page))
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(error => this.setState({ error }));
  };

  componentDidMount() {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  onSearchSubmit = e => {
    e.preventDefault();
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });

    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }
  };

  render() {
    const { searchTerm, results, searchKey, error } = this.state;

    const page =
      (results && results[searchKey] && results[searchKey].page) || 0;

    const list =
      (results && results[searchKey] && results[searchKey].hits) ||
      [];

    return (
      <div className="page">
        <h1>My Hacker Stories</h1>
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onChangeSearch}
            onSubmit={this.onSearchSubmit}
          >
            Search
          </Search>
        </div>

        {error ? (
          <p>Something went wrong.</p>
        ) : (
          <Table list={list} onDismiss={this.onDismiss} />
        )}

        <div className="interactions">
          <Button
            onClick={() =>
              this.fetchSearchTopStories(searchKey, page + 1)
            }
          >
            More
          </Button>
        </div>
      </div>
    );
  }
}

export default App;
