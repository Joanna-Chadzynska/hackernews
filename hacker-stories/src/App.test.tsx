import React from "react";
import renderer from "react-test-renderer";
import axios from "axios";

// import { render } from "@testing-library/react";
// import { Provider } from "react-redux";
// import { store } from "./app/store";
import App from "./App";
import { List } from "./components";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

// test suite
// describe("something truthy", () => {
// 	// test case
// 	it("true to be true", () => {
// 		// test assertion
// 		expect(true).toBe(true);
// 	});

// 	// test case
// 	it("false to be false", () => {
// 		// assertion
// 		expect(false).toBe(false);
// 	});
// });

describe("App", () => {
	it("succeeds fetching data with a list", async () => {
		const list = [
			{
				title: "React",
				url: "https://reactjs.org/",
				author: "Jordan Walke",
				num_comments: 3,
				points: 4,
				objectID: "0"
			},
			{
				title: "Redux",
				url: "https://redux.js.org",
				author: "Dan Abramov, Andrew Clark",
				num_comments: 2,
				points: 5,
				objectID: "1"
			}
		];

		const promise = Promise.resolve({
			data: {
				hits: list
			}
		});

		mockedAxios.get.mockImplementationOnce(() => promise);

		let component: any;
		await renderer.act(async () => {
			component = renderer.create(<App />);
		});

		expect(component.root.findByType(List).props.list).toEqual(list);
	});

	it("fails fetching data with a list", async () => {
		const promise = Promise.reject();

		mockedAxios.get.mockImplementationOnce(() => promise);

		let component: any;

		await renderer.act(async () => {
			component = renderer.create(<App />);
		});

		expect(component.root.findByType("p").props.children).toEqual(
			"Something went wrong ..."
		);
	});
});
