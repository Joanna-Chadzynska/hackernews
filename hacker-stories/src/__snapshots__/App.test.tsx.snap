// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem renders snapshot 1`] = `
<div
  className="sc-AxhCb XHhdY"
>
  <span
    className="sc-AxhUy iCwBwD"
    width="40%"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    className="sc-AxhUy iCuOPo"
    width="30%"
  >
    Jordan Walke
  </span>
  <span
    className="sc-AxhUy iCvnQi"
    width="10%"
  >
    3
  </span>
  <span
    className="sc-AxhUy iCvnQi"
    width="10%"
  >
    4
  </span>
  <span
    className="sc-AxhUy iCvnQi"
    width="10%"
  >
    <button
      className="sc-AxjAm sc-AxirZ eGSLNv button buttonSmall"
      onClick={[Function]}
      type="button"
    >
      <svg
        height="18px"
        width="18px"
      >
        logo.svg
      </svg>
    </button>
  </span>
</div>
`;
